grovel/grovel.lisp

```diff
@@ -179,68 +195,140 @@ int main(int argc, char**argv) {
-              (labels
-                  ((process-form (f)
-                     (case (form-kind f)
-                       (flag (warn "Groveler clause FLAG is deprecated, use CC-FLAGS instead.")))
-                     (case (form-kind f)
-                       (in-package
-                        (setf *package* (find-package (second f)))
-                        (push f forms))
-                       (progn
-                         ;; flatten progn forms
-                         (mapc #'process-form (rest f)))
-                       (t (push f forms)))))
-                (process-form form))))))
-   (let* ((forms (read-forms in))
-          (header-forms (remove-if-not #'header-form-p forms))
-          (body-forms (remove-if #'header-form-p forms)))
-     (write-string *header* out)
-     (dolist (form header-forms)
-       (process-grovel-form out form))
-     (write-string *prologue* out)
-     (dolist (form body-forms)
-       (process-grovel-form out form))
-     (write-string *postscript* out)
-     c-file)))
+(defun generate-c-file (c-file forms)
+  (with-standard-io-syntax
+    (let ((*print-readably* nil)
+          (*print-escape* t))
+      (with-open-file (out c-file :direction :output :if-exists :supersede)
+        (let* ((header-forms (remove-if-not #'header-form-p forms))
+               (body-forms (remove-if #'header-form-p forms)))
+          (write-string *header* out)
+          (dolist (form header-forms)
+            (process-grovel-form out form))
+          (write-string *prologue* out)
+          (dolist (form body-forms)
+            (process-grovel-form out form))
+          (write-string *postscript* out)
+          c-file)))))
 
 (defun tmp-lisp-file-name (defaults)
   (make-pathname :name (strcat (pathname-name defaults) ".grovel-tmp")
                  :type "lisp" :defaults defaults))
 
-
+
+(defun read-grovel-file (input-file)
+  (flet ((read-forms (s)
+           (do ((forms ()) (form (read s nil nil) (read s nil nil)))
+               ((null form) (nreverse forms))
+             (labels
+                 ((process-form (f)
+                    (case (form-kind f)
+                      (flag (warn "Groveler clause FLAG is deprecated, use CC-FLAGS instead.")))
+                    (case (form-kind f)
+                      (in-package
+                       (setf *package* (find-package (second f)))
+                       (push f forms))
+                      (progn
+                        ;; flatten progn forms
+                        (mapc #'process-form (rest f)))
+                      (t (push f forms)))))
+               (process-form form)))))
+    (with-open-file (in input-file :direction :input)
+      (with-cached-reader-conditionals (read-forms in)))))
+
+(defun touch-file (pathname)
+  (with-open-file (stream pathname :direction :output
+                          :if-exists :append
+                          :if-does-not-exist :create)))
```

Comment:
Ensures a file exists. Writes no data to it.

This is needed because of a limitation in asdf. If you tell it (in `#'perform`) that you will be making 3 files, you need to make those 3 files. There isnt a way to say 'actually we found out we didnt need them.

However this is the case when we load from cache. We don't need to create a the `c-file` as we have the `.so` file and the lisp file already. 

So we use this to make sure asdf doesnt complain to the user.